src/renderer/vt/state.cpp

```diff
@@ -187,35 +187,122 @@ VtEngine::VtEngine(_In_ wil::unique_hfile pipe,
+//      extensively by VtSequences.cpp
+// Arguments:
+// - estimatedSize: If you have a guess on how big the result will be, specify a non-zero
+//                  value here.
+//                  This is a performance optimization to skip counting the formatted string
+//                  prior to allocating memory and reformatting it again into the buffer.
+//                  If your guess is wrong, we will still fall back to the slow-way and
+//                  count then format.
+// - pFormat: pointer to format string to write to the pipe
+// - ...: a va_list of args to format the string with.
+// Return Value:
+// - S_OK, E_INVALIDARG for a invalid format string, or suitable HRESULT error
+//      from writing pipe.
+[[nodiscard]] HRESULT VtEngine::_WriteFormattedString(const size_t estimatedSize, const std::string* const pFormat, ...) noexcept
+{
+    // NOTE: pFormat is a pointer because varargs refuses to operate with a ref in that position
+    // NOTE: We're not using string_view because it doesn't guarantee null (which will be needed
+    //       later in the formatting method).
+
     va_list argList;
     va_start(argList, pFormat);
+    const auto hr = _WriteFormattedString(estimatedSize, pFormat, argList);
+    va_end(argList);
+    return hr;
+}
+
+// Method Description:
+// - Helper for calling _Write with a string for formatting a sequence. Used
+//      extensively by VtSequences.cpp
+// Arguments:
+// - estimatedSize: If you have a guess on how big the result will be, specify a non-zero
+//                  value here.
+//                  This is a performance optimization to skip counting the formatted string
+//                  prior to allocating memory and reformatting it again into the buffer.
+//                  If your guess is wrong, we will still fall back to the slow-way and
+//                  count then format.
+// - pFormat: pointer to format string to write to the pipe
+// - args: a va_list of args to format the string with.
+// Return Value:
+// - S_OK, E_INVALIDARG for a invalid format string, or suitable HRESULT error
+//      from writing pipe.
+[[nodiscard]] HRESULT VtEngine::_WriteFormattedString(const size_t estimatedSize, const std::string* const pFormat, va_list args) noexcept
+try
+{
+    // NOTE: pFormat is a pointer because varargs refuses to operate with a ref in that position
+    // NOTE: We're not using string_view because it doesn't guarantee null (which will be needed
+    //       later in the formatting method).
 
-    int cchNeeded = _scprintf(pFormat->c_str(), argList);
+    HRESULT hr = E_FAIL;
+
+    // We're going to hold onto our format string space across calls because
+    // the VT renderer will be formatting a LOT of strings and alloc/freeing them
+    // over and over is going to be way worse for perf than just holding some extra
+    // memory for formatting purposes.
+    static std::string formatString;

```

Comment:
strongly disprefer static here: I know there's only one VT renderer at a time in a process, but _man_ have we ever been mysteriously burned by these things before

Reply:
Oh, yeah, I'll hang it off the engine.